### PR TITLE
refactor(perception_utils): add test in object_classification

### PR DESCRIPTION
**Pull request by YoshiRi**
## Description

add test function for `object_classification.hpp`  in `common/perception_util`.

Here is test coverage.
image

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: common/perception_utils/test/src/test_object_classification.cpp [hunk 1]
```diff
@@ -61,3 +61,160 @@ TEST(object_classification, test_getHighestProbLabel)
...
+  {// True Case with uint8_t
+  std::vector<std::uint8_t> carlike_labels;
+  carlike_labels.push_back(ObjectClassification::BUS);
+  carlike_labels.push_back(ObjectClassification::CAR);
+  carlike_labels.push_back(ObjectClassification::TRAILER);
+  carlike_labels.push_back(ObjectClassification::TRUCK);
+  
+  for(auto label: carlike_labels){
+    EXPECT_TRUE(isCarLikeVehicle(label));
+  }
```
Comment (takayuki5168) [inline common/perception_utils/test/src/test_object_classification.cpp]: Is there any reason to make a variable of `carlike_labels`?

EXPECT_TRUE(isCarLikeVehicle(ObjectClassification::BUS)); looks better to me.

**Comment by takayuki5168** [comment]
@YoshiRi Thanks for your PR.
The title of the PR should be `refactor(package_name): `, and the package name is perception_util**s**.

refactor(perception_util)/

File: common/perception_utils/test/src/test_object_classification.cpp (hunk 1 above)
Comment (YoshiRi) [reply common/perception_utils/test/src/test_object_classification.cpp]: Thanks. I agree with you and fixed it.
At that time, I just thought it's better to use for loop so that I can prevent writing too many EXPECT_TRUE state.

File: common/perception_utils/test/src/test_object_classification.cpp [hunk 2]
```diff
@@ -61,3 +61,114 @@ TEST(object_classification, test_getHighestProbLabel)
...
+    classification.push_back(createObjectClassification(ObjectClassification::TRUCK, 0.8));
+    classification.push_back(createObjectClassification(ObjectClassification::BICYCLE, 0.7));
+    EXPECT_TRUE(isCarLikeVehicle(classification));
+  }
+
+  // False Case with object_classifications
+  {  // false case
+    std::vector<autoware_auto_perception_msgs::msg::ObjectClassification> classification;
+    classification.push_back(createObjectClassification(ObjectClassification::MOTORCYCLE, 0.8));
+    classification.push_back(createObjectClassification(ObjectClassification::BICYCLE, 0.8));
```
Comment (takayuki5168) [inline common/perception_utils/test/src/test_object_classification.cpp:139]: How about adding a test to see what will happen if the object classification includes car-like vehicle and not car-like vehicle with the same probability.
This will be the edge case for isCarLikeVehicle function.

File: common/perception_utils/test/src/test_object_classification.cpp [hunk 3]
```diff
@@ -61,3 +61,114 @@ TEST(object_classification, test_getHighestProbLabel)
     EXPECT_EQ(label, ObjectClassification::CAR);
```
Comment (takayuki5168) [inline common/perception_utils/test/src/test_object_classification.cpp:61]: image

pre-commit failed. Don't you know how to fix it on a github browser or locally with a command?

File: common/perception_utils/test/src/test_object_classification.cpp (hunk 2 above)
Comment (YoshiRi) [reply common/perception_utils/test/src/test_object_classification.cpp:139]: That makes sense. I'll do it.

File: common/perception_utils/test/src/test_object_classification.cpp (hunk 3 above)
Comment (YoshiRi) [reply common/perception_utils/test/src/test_object_classification.cpp:61]: It seems to be due to unnecessary line break in the code.
I fixed it in local and push later.

Comment (takayuki5168) [reply common/perception_utils/test/src/test_object_classification.cpp:61]: I meant there is a auto-fix command in both github browser and local command.

File: common/perception_utils/test/src/test_object_classification.cpp (hunk 2 above)
Comment (YoshiRi) [reply common/perception_utils/test/src/test_object_classification.cpp:139]: This edge case depends following function `getHighestProbLabel`, which extract highest-scored label from classification.

- This function do not considering multiple highest probability
- It returns an earlier label with maximum probability. (**It depends labels order!**)
- For example, if an objects has CAR and MOTORCYCLE label with the same probability,
  - `isCarLikeVehicle` return `true` when CAR label is added first
  - else it returns `false`

```cpp
inline std::uint8_t getHighestProbLabel(
  const std::vector<ObjectClassification> & object_classifications)
{
  std::uint8_t label = ObjectClassification::UNKNOWN;
  float highest_prob = 0.0;
  // TODO(Satoshi Tanaka): It might be simple if you use STL or range-v3.
  for (const auto & object_classification : object_classifications) {
    if (highest_prob < object_classification.probability) {
      highest_prob = object_classification.probability;
      label = object_classification.label;
    }
  }
  return label;
}
```

Comment (takayuki5168) [reply common/perception_utils/test/src/test_object_classification.cpp:138]: Got it. Even if the current implementation is not good, in order to show how it works with the edge case in the current implementation, adding this edge-case test seems useful.
What do you think?

File: common/perception_utils/test/src/test_object_classification.cpp (hunk 3 above)
Comment (YoshiRi) [reply common/perception_utils/test/src/test_object_classification.cpp:61]: Sorry, I forgot `pre-commit run -a` command before git push.
(But I still had to fix some error manually)

Comment (takayuki5168) [reply common/perception_utils/test/src/test_object_classification.cpp:61]: Got it.

`pre-commit install` which enables running pre-commit when committing every time is also useful.

File: common/perception_utils/test/src/test_object_classification.cpp (hunk 2 above)
Comment (YoshiRi) [reply common/perception_utils/test/src/test_object_classification.cpp:138]: Frankly, I think it is a rare situation that the label probabilities become the same since it is float value.
But adding edge case seems to be not harmful and it is easy to implement, I added some edge cases to this PR.

**Comment by takayuki5168** [comment]
@YoshiRi Could you resolve the conflict. Then this PR is ready to merge, right?

**Review by takayuki5168 — APPROVED** [review]
LGTM

**Review by scepter914 — APPROVED** [review]

**Comment by yukke42** [comment]
@YoshiRi Have you forgotten to merge?
